### ISSUE TITLE
Multi-GPU training but get lower mAP

### ISSUE THREAD

**Issue by Joker316701882** [ultralytics/yolov5#2843]
## ❔Question

Hello @glenn-jocher , I trained yolov5 medium and large on 8 GPUs with the following command but got the following final mAP::
**v5m**: `python -m torch.distributed.launch --nproc_per_node 8 train.py --data data/coco.yaml --cfg yolov5m.yaml --weights '' --batch-size 64`
![image](https://user-images.githubusercontent.com/18679821/115189304-d8971880-a118-11eb-8aff-d6cbcc034696.png)
**v5l**: `python -m torch.distributed.launch --nproc_per_node 8 train.py --data data/coco.yaml --cfg yolov5l.yaml --weights '' --batch-size 64`
![image](https://user-images.githubusercontent.com/18679821/115189183-ad142e00-a118-11eb-8e98-3110b686ca2f.png)

Both performances are lower than the reported in README.md. You mentioned in some issues that **sync-bn is better being used when training with multi-gpus** for better batch statistics. However, in your batch size study https://github.com/ultralytics/yolov5/discussions/2452, it seems the batch size (batch statistics) does not matter. In other words, I don't think sync-bn could help my case given the information in https://github.com/ultralytics/yolov5/discussions/2452.

Do you have any other ideas that might cause the performance drop when training with multi-gpu compared to single-gpu?

**Comment by Ronald-Kray** [comment]
@Joker316701882 
Hey I have questions.
How could your get this result? could you share the code with me, please? 
![image](https://user-images.githubusercontent.com/76469976/115277892-c2707300-a17f-11eb-88b7-7488b0bdeae8.png)

**Comment by glenn-jocher** [comment]
@Joker316701882 hey buddy. So there are two items here you may want to investigate. The first is that the current results for P5 models like the 5m and 5l models you are displaying were produced on single-GPU. Yes the batch-size study shows good independence of batch size with final mAP but I don't think you can make any assumptions on the multi-GPU side, so first things first you should re-run with single-GPU and/or with multi-GPU and --sync enabled to evaluate the difference. I've been meaning to run a --sync study but have not had the opportunity to yet, if you get to these results first and are willing to share then perhaps we can use your runs for this study!

The second is that the larger models overtrain more easily and benefit more from additional augmentation, so most of the current models were trained at scale: 0.9 rather than scale: 0.5, in following with the EfficientDet appendix results at https://arxiv.org/pdf/1911.09070.pdf

**Comment by Joker316701882** [comment]
@glenn-jocher Hey man. Thank you for your reply, I already run with the --sync-bn flag and will update the result here soon. It will be interesting to conduct more experiments to compare the behaviors of single-GPU and multi-GPU w/ and w/o --sync-bn. I will run the experiments as long as we have idle devices. About the large scale jitter, it does allow our internal one-stage detector to train with lower lrf (lrf=0.01). Thank you for your valuable suggestions and let's keep in touch.

**Comment by Joker316701882** [comment]
@Ronald-Kray Hey. I think you can get the above results as long as you finish your training with the following command:

`python -m torch.distributed.launch --nproc_per_node 8 train.py --data data/coco.yaml --cfg yolov5l.yaml --weights '' --batch-size 64`

I use 8 2080Tis.

**Comment by Joker316701882** [comment]
@glenn-jocher Hi. I tried the following command which is exactly the same as in README.md for training a large model:
`python train.py --data coco.yaml --cfg yolov5l.yaml --weights '' --batch-size 24`

Relevant information:
Pytorch 1.8.1, single Tesla V100 GPU.

But got the following best mAP:
![image](https://user-images.githubusercontent.com/18679821/116056711-aa817d80-a6b0-11eb-855e-d731161fb354.png)
The mAP on the COCO _testdev_ might be slightly better than this. But it's still lower than **48.1**mAP reported as in README. Any ideas? 
I also notice that in the above command, **multiscale training** is not used, should I try to turn on the multiscale flag? (Although scale jittering is already adopted in the **mosaic** implementation)

**Comment by glenn-jocher** [comment]
@Joker316701882 you might want to increase your scale jitter from 0.5 to 0.9 for larger models to delay overfitting.

**Comment by Joker316701882** [comment]
@glenn-jocher Sure, I'll try it. No offense but should there be a clear way/setting to reproduce the reported performance?

**Comment by Joker316701882** [comment]
@glenn-jocher I increase the scale jitter from 0.5 to 0.9 and got the following result for yolov5 l:
![image](https://user-images.githubusercontent.com/18679821/117932194-abfaa900-b332-11eb-856a-a28d96f69131.png)

The mAP on _testdev_ is usually 0.2~0.3 better than on _val set_, so I think the performance reported in README is reproducible. Thank you for your information.

**Comment by glenn-jocher** [comment]
@Joker316701882 yes this looks about right. The limiting factor in producing better mAP on larger models like 5l, 5x seems to always be overfitting on the objectness and cls losses, but mostly the objectness loss. You can improve the above results slightly by reducing the two gains, or the two positive weights, or both. 

But on the contrary these steps will hurt smaller models like 5s and 5m. For the 5x models we train with these hyps that slightly reduce the two BCEloss gains.

```yaml
lr0: 0.01  # initial learning rate (SGD=1E-2, Adam=1E-3)
lrf: 0.2  # final OneCycleLR learning rate (lr0 * lrf)
momentum: 0.937  # SGD momentum/Adam beta1
weight_decay: 0.0005  # optimizer weight decay 5e-4
warmup_epochs: 3.0  # warmup epochs (fractions ok)
warmup_momentum: 0.8  # warmup initial momentum
warmup_bias_lr: 0.1  # warmup initial bias lr
box: 0.05  # box loss gain
cls: 0.3  # cls loss gain
cls_pw: 1.0  # cls BCELoss positive_weight
obj: 0.7  # obj loss gain (scale with pixels)
obj_pw: 1.0  # obj BCELoss positive_weight
iou_t: 0.20  # IoU training threshold
anchor_t: 4.0  # anchor-multiple threshold
anchors: 3  # anchors per output layer (0 to ignore)
fl_gamma: 0.0  # focal loss gamma (efficientDet default gamma=1.5)
hsv_h: 0.015  # image HSV-Hue augmentation (fraction)
hsv_s: 0.7  # image HSV-Saturation augmentation (fraction)
hsv_v: 0.4  # image HSV-Value augmentation (fraction)
degrees: 0.0  # image rotation (+/- deg)
translate: 0.1  # image translation (+/- fraction)
scale: 0.9  # image scale (+/- gain)
shear: 0.0  # image shear (+/- deg)
perspective: 0.0  # image perspective (+/- fraction), range 0-0.001
flipud: 0.0  # image flip up-down (probability)
fliplr: 0.5  # image flip left-right (probability)
mosaic: 1.0  # image mosaic (probability)
mixup: 0.0  # image mixup (probability)
```

**Comment by glenn-jocher** [comment]
@Joker316701882 also its unclear if --sync helps or not. In the early stages of training it helps a lot. For final mAP I haven't seen great evidence that it makes an improvement.
